File: src/cpp/Source/Chunks/Common/ShaderVersion.cpp
#include "PCH.h"
#include "ShaderVersion.h"

#include "Decoder.h"

using namespace std;
using namespace SlimShader;

ShaderVersion ShaderVersion::ParseShex(BytecodeReader& reader)
{
	auto versionToken = reader.ReadUInt32();
	
	ShaderVersion result;
	result._minorVersion = DecodeValue<uint8_t>(versionToken, 0, 3);
	result._majorVersion = DecodeValue<uint8_t>(versionToken, 4, 7);
	result._programType = DecodeValue<ProgramType>(versionToken, 16, 31);
	return result;
}

ShaderVersion ShaderVersion::ParseRdef(BytecodeReader& reader)
{
	auto target = reader.ReadUInt32();

	auto programTypeValue = DecodeValue<uint16_t>(target, 16, 31);
	ProgramType programType;
	switch (programTypeValue)
	{
	case 0xFFFF:
		programType = ProgramType::PixelShader;
		break;
	case 0xFFFE:
		programType = ProgramType::VertexShader;
		break;
	case 0x4853:
		programType = ProgramType::HullShader;
		break;
	case 0x4753:
		programType = ProgramType::GeometryShader;
		break;
	case 0x4453:
		programType = ProgramType::DomainShader;
		break;
	case 0x4353:
		programType = ProgramType::ComputeShader;
		break;
	default:
		throw runtime_error("Unknown program type: " + to_string(programTypeValue));
	}

	ShaderVersion result;
	result._majorVersion = DecodeValue<uint8_t>(target, 8, 15);
	result._minorVersion = DecodeValue<uint8_t>(target, 0, 7);
	result._programType = programType;
	return result;
}

uint8_t ShaderVersion::GetMajorVersion() const { return _majorVersion; }
uint8_t ShaderVersion::GetMinorVersion() const { return _minorVersion; }
ProgramType ShaderVersion::GetProgramType() const { return _programType; }

File: src/SlimShader/Chunks/Common/ShaderVersion.cs
using SlimShader.Util;

namespace SlimShader.Chunks.Common
{
	/// <summary>
	/// Version Token (VerTok)
	/// [07:00] minor version number (0-255)
	/// [15:08] major version number (0-255)
	/// [31:16] D3D10_SB_TOKENIZED_PROGRAM_TYPE
	/// </summary>
	public class ShaderVersion
	{
		public byte MajorVersion { get; private set; }
		public byte MinorVersion { get; private set; }
		public ProgramType ProgramType { get; private set; }

		public static ShaderVersion ParseShex(BytecodeReader reader)
		{
			uint versionToken = reader.ReadUInt32();
			return new ShaderVersion
			{
				MinorVersion = versionToken.DecodeValue<byte>(0, 3),
				MajorVersion = versionToken.DecodeValue<byte>(4, 7),
				ProgramType = versionToken.DecodeValue<ProgramType>(16, 31)
			};
		}

		public static ShaderVersion ParseRdef(BytecodeReader reader)
		{
			uint target = reader.ReadUInt32();

			var programTypeValue = target.DecodeValue<ushort>(16, 31);
			ProgramType programType;
			switch (programTypeValue)
			{
				case 0xFFFF:
					programType = ProgramType.PixelShader;
					break;
				case 0xFFFE:
					programType = ProgramType.VertexShader;
					break;
				case 0x4853:
					programType = ProgramType.HullShader;
					break;
				case 0x4753:
					programType = ProgramType.GeometryShader;
					break;
				case 0x4453:
					programType = ProgramType.DomainShader;
					break;
				case 0x4353:
					programType = ProgramType.ComputeShader;
					break;
				default:
					throw new ParseException(string.Format("Unknown program type: 0x{0:X}", programTypeValue));
			}

			return new ShaderVersion
			{
				MajorVersion = target.DecodeValue<byte>(8, 15),
				MinorVersion = target.DecodeValue<byte>(0, 7),
				ProgramType = programType
			};
		}
	}
}